### PR TITLE
Feature lwdev 6836 black list

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: AutorestClient/BlockchainCashoutPreconditionsCheckAPI.cs [hunk 1]
```diff
@@ -0,0 +1,1480 @@
+// <auto-generated>
```
Comment (KonstantinRyazantsev) [inline AutorestClient/BlockchainCashoutPreconditionsCheckAPI.cs]: Seems that this file along another in the /AutorestClient folder should be removed

File: client/Lykke.Service.BlockchainCashoutPreconditionsCheck.Client/IBlockchainCashoutPreconditionsCheckClient.cs [hunk 2]
```diff
@@ -7,6 +8,47 @@ namespace Lykke.Service.BlockchainCashoutPreconditionsCheck.Client
...
+        /// <exception cref="Exception">Is thrown on wrong usage of service.</exception>
         Task<(bool isAllowed, IEnumerable<ValidationErrorResponse>)> ValidateCashoutAsync(CashoutValidateModel validateCashoutModel);
+
+        /// <summary>
+        /// Add new address to the specific blockchain type black list
+        /// </summary>
+        /// <param name="blackListModel"></param>
+        /// <returns></returns>
+        /// <exception cref="Exception">Is thrown on wrong usage of service.</exception>
+        Task CreateBlackListAsync(BlackListModel blackListModel);
```
Comment (KonstantinRyazantsev) [inline client/Lykke.Service.BlockchainCashoutPreconditionsCheck.Client/IBlockchainCashoutPreconditionsCheckClient.cs]: AddToBlackList will be suitable name

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+using Lykke.AzureStorage.Tables;
+using Lykke.Service.BlockchainCashoutPreconditionsCheck.Core.Domain.Validation;
+
+namespace Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories.Entities
+{
+    internal class BlackListEntity : AzureTableEntity
+    {
+        #region Fields
+
+        // ReSharper disable MemberCanBePrivate.Global
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs]: Better to use ```[UsedImplicitly(ImplicitUseKindFlags.Assign)]``` on the property, this takes a chance to perform more Resharper checks

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        #region Keys
+
+        public static string GetPartitionKey(string blockchainIntegrationLayerId)
+        {
+            return blockchainIntegrationLayerId;
+        }
+
+        public static string GetRowKey(string blockedAddress)
+        {
+            return blockedAddress.ToLower();
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs:39]: Seems, that we should to do ToLower() only if IsCaseSensetive

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+        public static BlackListEntity FromDomain(BlackListModel model)
+        {
+            return new BlackListEntity
+            {
+                PartitionKey = GetPartitionKey(model.BlockchainType),
+                RowKey = GetRowKey(model.BlockedAddress),
+                BlockchainIntegrationLayerId = model.BlockchainType,
+                BlockedAddress = model.BlockedAddress,
+                BlockedAddressLowCase = model.BlockedAddress?.ToLower(),
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs:55]: Consider to store address in the single field, but make ToLower() only when !IsCaseSensitive

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        // ReSharper restore MemberCanBePrivate.Global
+
+        #endregion
+
+
+        #region Keys
+
+        public static string GetPartitionKey(string blockchainIntegrationLayerId)
+        {
+            return blockchainIntegrationLayerId;
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs:34]: Not scalable PK. What if we need to block a lot of addresses within single blockchain?

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        public static BlackListEntity FromDomain(BlackListModel model)
+        {
+            return new BlackListEntity
+            {
+                PartitionKey = GetPartitionKey(model.BlockchainType),
+                RowKey = GetRowKey(model.BlockedAddress),
+                BlockchainIntegrationLayerId = model.BlockchainType,
+                BlockedAddress = model.BlockedAddress,
+                BlockedAddressLowCase = model.BlockedAddress?.ToLower(),
+                IsCaseSensitiv = model.IsCaseSensitive,
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs]: Typo - Sensitiv**e**

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/ValidationErrorType.cs [hunk 8]
```diff
@@ -1,12 +1,13 @@
 ﻿namespace Lykke.Service.BlockchainCashoutPreconditionsCheck.Core.Domain.Validations
 {
     public enum ValidationErrorType
-    { 
+    {
         None,
         AddressIsNotValid,
-        FieldNotValid,
+        FieldIsNotValid,
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/ValidationErrorType.cs:7]: This is API breaking change (if StringEnumConverter is enabled for Json.Net). How we will deploy the feature? If StringEnumConverter is enabled, then seems that we could deploy LW API first, if we preserve old enum value (marked as obsolete) and add new one.

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Requests/CheckCashoutValidityModel.cs [hunk 9]
```diff
@@ -20,5 +20,8 @@ public class CheckCashoutValidityModel
         [Required]
         [DataMember]
         public string DestinationAddress { get; set; }
+
+        [Required]
+        public string DestinationAddressBase { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Requests/CheckCashoutValidityModel.cs]: This API breaking change. We should deploy LW API first to support this field, but are another changes permit this scenario?

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Responses/BlackListResponse.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        public string BlockchainType { get; set; }
+
+        [DataMember]
+        public string BlockedAddress { get; set; }
+
+        [DataMember]
+        public bool IsCaseSensitive { get; set; }
+    }
+
+    public class BlackListEnumerationResponse
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Responses/BlackListResponse.cs]: Multiple classes in the single file

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            var blackList = await _blackListRepository.TryGetAsync(blockchainType, blockedAddress);
+
+            if (blackList == null)
+            {
+                return false;
+            }
+
+            bool isBlocked = false;
+
+            isBlocked = (blackList.IsCaseSensitive && blockedAddress == blackList.BlockedAddress) ||
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs:36]: We'll don't need to do this, if RK will be stored in appropriate casing

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+
+        private async Task<IBlockchainApiClient> ThrowOnNotSupportedBlockchainType(string blockchainType)
+        {
+            IBlockchainApiClient blockchainClient;
+
+            try
+            {
+                blockchainClient = _blockchainApiClientProvider.Get(blockchainType); //throws
+            }
+            catch (Exception e)
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs]: Better to catch more specific exception, that BlockchainApiClientProvider throws when blockchain is not found, or even add TryGetValue to the provider

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services.csproj [hunk 13]
```diff
@@ -16,7 +16,14 @@
   </ItemGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories\Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories.csproj" />
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services.csproj]: Services should not depends on repositories

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services.csproj [hunk 14]
```diff
@@ -16,7 +16,14 @@
...
     <ProjectReference Include="..\Lykke.Service.BlockchainCashoutPreconditionsCheck.Core\Lykke.Service.BlockchainCashoutPreconditionsCheck.Core.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Reference Include="System.Threading.Tasks">
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services.csproj]: Strange reference

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/CashoutModel.cs [hunk 15]
```diff
@@ -7,7 +7,11 @@ namespace Lykke.Service.BlockchainCashoutPreconditionsCheck.Core.Domain.Validati
     public class CashoutModel
     {
         public string AssetId { get; set; }
+
         public decimal Volume { get; set; }
+
         public string DestinationAddress { get; set; }
+
+        public string DestinationAddressBase { get; set; }
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/CashoutModel.cs]: This field really should be added to the contract? this service could use BlockchainWallets API to check if given blockchain support address extensions and extract base and extension address parts. This should be already done in the BlockchainCashoutPreconditionsCheck in the LWDEV-6879

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+using Lykke.Service.BlockchainCashoutPreconditionsCheck.Models.Responses;
+using Microsoft.AspNetCore.Mvc;
+using Swashbuckle.AspNetCore.SwaggerGen;
+
+namespace Lykke.Service.BlockchainCashoutPreconditionsCheck.Controllers
+{
+    [Route("api/[controller]")]
+    public class BlackListController : Controller
+    {
+        private readonly char[] _charactersToTrim = new char[] {' ', '\t'};
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: could be made static

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+        public BlackListController(IBlackListService blackListService)
+        {
+            _blackListService = blackListService;
+        }
+
+        /// <summary>
+        /// is address black listed
+        /// </summary>
+        /// <returns></returns>
+        [HttpGet("is-blocked/{blockchainType}/{blockedAddress}")]
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: {blockchainType}/{blockedAddress}/is-blocked would be more RESTful

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+        [ProducesResponseType(typeof(ErrorResponse), (int)HttpStatusCode.InternalServerError)]
+        [ProducesResponseType(typeof(ErrorResponse), (int)HttpStatusCode.BadRequest)]
+        public async Task<IActionResult> AddBlockedAddressAsync([FromBody] AddBlackListModel request)
+        {
+            string blockedAddress = Trim(request.BlockedAddress);
+
+            BlackListModel model = new BlackListModel()
+            {
+                BlockedAddress = blockedAddress,
+                BlockedAddressLowCase = blockedAddress.ToLower(),
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: ToLower shouldn't be here, even if BlockedAddressLowCase will be preserved. Move this to iether domain model - pass parameters via ctor or static factory method and initialize BlockedAddressLowCase there, on in the service - pass parameters to the SaveAsync and make new BlackListModel there.

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+        [ProducesResponseType(typeof(ErrorResponse), (int)HttpStatusCode.InternalServerError)]
+        [ProducesResponseType(typeof(ErrorResponse), (int)HttpStatusCode.BadRequest)]
+        public async Task<IActionResult> UpdateBlockedAddressAsync([FromBody] AddBlackListModel request)
+        {
+            string blockedAddress = Trim(request.BlockedAddress);
+
+            BlackListModel model = new BlackListModel()
+            {
+                BlockedAddress = blockedAddress,
+                BlockedAddressLowCase = blockedAddress.ToLower(),
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: The same thing as above

File: AutorestClient/BlockchainCashoutPreconditionsCheckAPI.cs (hunk 1 above)
Comment (OttoVT) [reply AutorestClient/BlockchainCashoutPreconditionsCheckAPI.cs]: it is used inside BlockchainCashoutPreconditionsCheckClient

File: client/Lykke.Service.BlockchainCashoutPreconditionsCheck.Client/IBlockchainCashoutPreconditionsCheckClient.cs (hunk 2 above)
Comment (OttoVT) [reply client/Lykke.Service.BlockchainCashoutPreconditionsCheck.Client/IBlockchainCashoutPreconditionsCheckClient.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs (hunk 3 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs (hunk 4 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs:39]: Nope, IsCaseSensetive is a flag for business logic, We store all addresses in lower case

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs (hunk 5 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs:55]: There should be no logic for storing items.

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs (hunk 7 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.AzureRepositories/Entities/BlackListEntity.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/ValidationErrorType.cs (hunk 8 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/ValidationErrorType.cs:7]: This is not a breaking change.
Look at Lykke.Service.BlockchainCashoutPreconditionsCheck.Models.Responses.ValidationErrorType

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Requests/CheckCashoutValidityModel.cs (hunk 9 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Requests/CheckCashoutValidityModel.cs]: Made as not required

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Responses/BlackListResponse.cs (hunk 10 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Responses/BlackListResponse.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs (hunk 12 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/BlackListService.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services.csproj (hunk 13 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services/Lykke.Service.BlockchainCashoutPreconditionsCheck.Services.csproj]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs (hunk 16 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs (hunk 17 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/CashoutModel.cs (hunk 15 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Core/Domain/Validation/CashoutModel.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs (hunk 18 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs (hunk 19 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck/Controllers/BlackListController.cs]: fixed

File: src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Requests/CheckCashoutValidityModel.cs (hunk 9 above)
Comment (OttoVT) [reply src/Lykke.Service.BlockchainCashoutPreconditionsCheck.Models/Requests/CheckCashoutValidityModel.cs]: removed

**Review by KonstantinRyazantsev — APPROVED** [review]

**Pull request merged**